### ISSUE TITLE
Writeup: Resolving differences between CS2030 notes and Oracle's JVM spec

### ISSUE THREAD

**Issue by JuliaPoo** [nus-cs2030/2223-s1#297]
# Metadata

While reading the notes I had some discrepancies within the notes' description of the Java Virtual Machine (JVM), in particular Stack vs Heap, and my internal conception of what's happening, and attempted to resolve my understanding by consulting the [JVM spec](https://docs.oracle.com/javase/specs/jvms/se19/html/) and disassembling the java bytecode. Upon doing so, I noticed significant discrepancies between what the notes present and the spec, and I attempt to resolve them here. I might be adding more if I have time and have more questions.

If there are any errors I made do point them out, I have no experience with the JVM prior to this.

Clarifications:
1. Am using the spec for SE19
2. When I say 'stack' and 'heap' and other objects, I'll be referring to these objects in the context of the JVM and not the analogue in native execution (E.g. **JVM stack** vs Native stack, **JVM heap** vs Native heap). This distinction is important cuz they don't necessarily coincide. JVM can allocate (jvm) stack frames in the native heap.
3. When talking abt JVM execution, I'll situation of no native execution (E.g., no jitting, no entering native methods)
4. Only single-threaded execution. 

## Preliminaries

- JVM is the object that executes java. Simplistically, Java source is compiled into [Java Bytecode](https://en.wikipedia.org/wiki/List_of_Java_bytecode_instructions), which the JVM then executes. I consult the spec and the bytecode for this writeup.
 
## Stack VS Heap: The notes VS the Spec

![Notes 1](https://i.imgur.com/k8fLJVW.jpg)

### The Notes

The notes (Lecture 1) describe the stack as a "LIFO (Last In First Out) stack for storing activation records of method calls". In it, we see the stack containing the local variables for each nested method call. This conception aligns well with my intuition for the C-stack. However, it starts to break down when it comes to questions like "When exiting a method call (e.g., `return`), how does JVM know where to continue execution?", or "How does JVM know how much to pop from the stack upon exiting a method".

### The Spec

The Spec _kinda_ agrees with the note's conception of the stack: JVM does have a LIFO stack that tracks method calls. However, each element in the stack isn't a local variable: Each element is a _frame_.

A _frame_ is added when JVM enters a method, and removed when it returns from a method. Note, **ONE** _frame_ is added to the stack **PER** method entered. Each _frame_, then contains a _Local Variable Array_ and an _Operand Stack_. So at any point in time, a JVM thread manages at least **TWO** stacks: The stack that contains the frames (ill be calling this the _Call Stack_), and the current frame's operand stack. Note that the local variables of the current frame aren't in a stack: They are in an array.

The spec describes JVM as both a [Register Machine](https://en.wikipedia.org/wiki/Register_machine) and a [Stack Machine](https://en.wikipedia.org/wiki/Stack_machine). 

When JVM enters a method, a frame is created and the _Local Variable Array_ is populated with the method's arguments (E.g., calling `f(a=1, b=2)` adds `(int a,1), (int b, 2)` key-value pairs to the Local Variable Array). The _Local Variable Array_ is also appended to during runtime (E.g., when executing `int a = 1`, and `(int a, 1)` key-value pair is added to the _Local Variable Array_). The operand stack, on the other hand, performs intermediate operations in computing the value of the local variables/return value. (E.g., `a = 1+3`, JVM uses the operand stack to compute `1+3 = 4` and stores the result in the _Local Variable Array_ entry for `a`). 

Aside: In class methods that have the `this` variable. Within JVM, the `this` variable, a reference to its class, is passed in as the first argument in the method, so the first element of the Local Variable Array in such a method is the `(this, <reference in heap>)` key-value pair.

The operand stack and local variable array can, of course, store references to objects in the heap as well, and the operand stack can store references to the local variable array. Do note that these are only in the current frame. To my knowledge, there isn't a way for a current frame to directly access the operand stack/local variable array of another frame. 

When JVM exits the method, the return value is passed to the previous frame _somehow_ (idk how) and the current frame is removed. Execution continues in the previous frame.

Refer to [Here](https://cscircles.cemc.uwaterloo.ca/java_visualize/) for an incomplete visualisation of what I mean (The link visualises the Call Stack and the Local Variable Array but not the Operand Stack, and doesn't account for everything, like inner classes).

### Attempt to resolve

I feel like the stack as described in the notes combines the Local Variable Array of all frames on the call stack into one big stack, with the order on the stack being the order in which the variables are created? This is a _very_ liberal interpretation of the stack and I have no idea how to resolve this discrepancy. 

## Accessing variables that isn't in the stack

Okay, so each frame's code can access the Local Variable Array, which includes method arguments. But as we know, code within a method can access way more than that. E.g.,

```java
class B {
  void f() {}
}

public class Main {
  public static void main(String[] args) {
    B b = new B();
    b.f();
  }
}
```

Within `Main.main`, the symbols `B` and `B.f` are not in `Main.main`'s Local Variable Array, and yet they can be accessed.

The solution is in the _Constant Pool_. Each frame has its own _Constant Pool_ that contains all references to symbols that the frame _should_ be able to access, but isn't a local variable or an argument. E.g., for the example above we have:

```
Main.main Constant Pool:
     #1 = Methodref          #2.#3          // java/lang/Object."<init>":()V
     #2 = Class              #4             // java/lang/Object
     #3 = NameAndType        #5:#6          // "<init>":()V
     #4 = Utf8               java/lang/Object
     #5 = Utf8               <init>
     #6 = Utf8               ()V
     #7 = Class              #8             // B
     #8 = Utf8               B
     #9 = Methodref          #7.#3          // B."<init>":()V
    #10 = Methodref          #7.#11         // B.f:()V
    #11 = NameAndType        #12:#6         // f:()V
    #12 = Utf8               f
    #13 = Class              #14            // Main
    #14 = Utf8               Main
    #15 = Utf8               Code
    #16 = Utf8               LineNumberTable
    #17 = Utf8               main
    #18 = Utf8               ([Ljava/lang/String;)V
    #19 = Utf8               SourceFile
    #20 = Utf8               Main.java

Main.main bytecode:
    0: new           #7                  // class B
    3: dup
    4: invokespecial #9                  // Method B."<init>":()V
    7: astore_1
    8: aload_1
    9: invokevirtual #10                 // Method B.f:()V
    12: return
```

Do note that in the _Constant Pool_, not all symbols have been 'initialised'. E.g., In the above, the spec totally allows `#7` (a reference to  class `B`) to reference _nothing_ UNTIL `Main.main` needs it, at which point the file `B.class` will be loaded into memory and `#7` is made to reference class `B`.

## Inner Classes: Heap layout explanation

TODO: How inner classes are compiled

## TODO if have time and motivation

1. How `return` works: How the JVM knows where to continue execution upon exiting a method and how is the return value passed to the previous frame.

## Resources

1. [JVM Spec](https://docs.oracle.com/javase/specs/jvms/se19/html/)
2. [Java Bytecode](https://en.wikipedia.org/wiki/List_of_Java_bytecode_instructions)
5. [Incomplete JVM Stack Frame Visualiser](https://cscircles.cemc.uwaterloo.ca/java_visualize/)
6. [Javap: Disassembler Reference API](https://docs.oracle.com/javase/7/docs/technotes/tools/windows/javap.html)

**Comment by JuliaPoo** [comment]
I'll be writing up on how the JVM deals with inner classes. The simulation I linked is incomplete and hence ur screenshot of it is also incomplete. From what I gather, the note's heap layout for the instance of abstract class `A` is accurate. The reason why it assumes such a layout however, is never mentioned in the notes and I'm guessing we are _somehow_ meant to infer why.

**Comment by eLTehh** [comment]
Yeah, that's a good point tbh. Also I double checked SE17 (the one we're using) vs. SE19 (your reference) docs, and both are identical with regards to the JVM stack.

My thoughts:

From the recit, it does seem like they're drawing the stack in 
1. order of frames introduced
2. order of local variables introduced, for each frame. additionally, these are actually the pointers and not *actually* the variables themselves (as they are stored in some other array of constants)

I do agree that mentioning frames (and maybe the operand stack) may have been a clearer way to illustrate the JVM stack. Would like to know prof's thoughts on this. Maybe because 2040 is separate from 2030 that it's not gone into detail so as to not confuse students? Idk. Frames could be a nice resource for folks who want to do further reading in the future.
